Bindings/GLFW.hsc

```diff
@@ -466,3 +469,4 @@ deriving instance Data     C'GLFWcursor
 
 #ccall glfwVulkanSupported , IO CInt
 #ccall glfwGetRequiredInstanceExtensions , Ptr CUInt -> IO (Ptr CString)
+#ccall glfwCreateWindowSurface , Ptr Void ->  Ptr <GLFWwindow> -> Ptr Void -> Ptr Word64 -> IO Int32
```

Comment:
Personally, I think `Ptr ()` is better for void pointers. Also, why not make all of these pointers `Ptr ()` to avoid the new `Data` imports?

Reply:
Personally, I like `Ptr Void` because `Void` does not have a `Storable` instance, and the type system tells you explicitly that you should not "dereference" this pointer; I also do no see a reason to avoid importing `Data.Void` :)

But the real reason for this decision was to match types from the vulkan bindings (which is not necessary, but handy).

Reply:
So, I totally think that the lastest GLFW should have support for vulkan, and I think this PR is on the right track. It should be implemented without outside influence, however. (Why should it match your vulkan bindings and not e.g. [vulkan](https://hackage.haskell.org/package/vulkan)?)

There are already instances in the library where we use `Ptr ()` (such as for `glfwGetWindowUserPointer`). Being inconsistent is worse than supporting an interface to another library.

Also, at the end of the day, `Ptr Void` doesn't really give any indication as to what the types should be. It'd be better to declare new void data types without constructors that you can use to communicate what the associated pointers should be. Something like:

    data VkInstance
    data VkAllocationCallbacks
    data VkSurfaceKHR
    #ccall glfwCreateWindowSurface , Ptr VkInstance ->  Ptr <GLFWwindow> -> Ptr VkAllocationCallbacks -> Ptr VkSurfaceKHR -> IO CInt

This wouldn't change the way your proposal interfaces with GLFW-b, and still has all of the properties you laid out for `Void` (no Storable instance and the type system prevents dereferencing).

Reply:
If we are discussing `Ptr ()` vs `Ptr Void`, I strongly believe the latter is better at representing `*void`. And both [vulkan](https://hackage.haskell.org/package/vulkan-1.7.0.0) and [vulkan-api](https://hackage.haskell.org/package/vulkan-api) agree on that (e.g. [look at `pNext` field of any struct in `vulkan`](https://hackage.haskell.org/package/vulkan-1.7.0.0/docs/Graphics-Vulkan-Buffer.html#t:VkBufferCreateInfo)). However you have a strong argument about consistency, so I totally agree to change the types if you are going to merge it :)

Having dedicated empty types seemed to me the best idea in the beginning, but I decided not to do that because that would introduce type dependencies on the package. For example, `VkInstance` is used a lot across vulkan api, and it would be ridiculous to have it declared outside of the library.

However, all of this does not make a lot of difference from the user perspective if my [PR to `GLFW-b`](https://github.com/bsl/GLFW-b/pull/76) gets accepted, because the [`createWindowSurface`](https://github.com/bsl/GLFW-b/pull/76/files#diff-dcee3bce55b07e53ba6c8fbbaa1c069fR1269) there uses type variables to be compatible with any vulkan bindings.

Reply:
OK -- I think your experience with Vulkan is great and super valuable. How about this:

These bindings are supposed to be super low level, such that higher level libraries like GLFW-b can build on top of them. So ideally, they're going to mirror the C code as closely as possible. With that in mind, how about we use the parameterized `Ptr` types you have in GLFW-b here instead:

```
#ccall glfwCreateWindowSurface , Ptr vkInstance -> Ptr <GLFWwindow> -> Ptr vkAllocationCallbacks -> Ptr vkSurfaceKHR -> IO CInt
```

That way anyone who uses the *low-level bindings* won't need to worry about what the types are of the higher level libraries. It wouldn't introduce type dependencies, etc. My take is that people using this library directly are already fairly comfortable with C, so hand-holding via type constraints is likely only going to get in people's way. Thoughts?

Reply:
I just tried this and it works -- GHCi reports that the following symbol exists:

```
p'glfwCreateWindowSurface ::
  GHC.Ptr.FunPtr
    (Ptr vkInstance
     -> Ptr C'GLFWwindow
     -> Ptr vkAllocationCallbacks
     -> Ptr vkSurfaceKHR
     -> IO Foreign.C.Types.CInt)
```

However, it wasn't able to find the non-function-pointer version, presumably due to the lack of `#define` enabling Vulkan.